File: backend/src/workers/telegramWorker.ts
import { TelegramClient } from 'telegram';
import { StringSession } from 'telegram/sessions';
import { NewMessage, NewMessageEvent } from 'telegram/events';
import { parseTelegramText } from '../services/parser';
import prisma from '../db';
import { Server } from 'socket.io';
import { processExternalThreat } from '../services/aggregatorService';

const CHANNELS = [
  'vanek_nikolaev', 
  'monitor', 
  'kievreal1', 
  'operativnoZSU',
  'insiderUKR',
  'smolii_ukraine',
  'kyiv_golovne',
  'war_monitor',
  'monitor_ukraine',
  'radar_ua_top',
  'kyiv_vanek',
  'nebo_raketa',
  'sectorv666',
  'deraketaua',
  'eradarrrua',
  'kherson_non_drone',
  'rozvidkaneba',
  'kudy_letyt',
  'nablydatel_dozor',
  'krolevetsnews',
  'place_kharkiv',
  'ukraineradar_24_7',
  'kievmap',
  'truexanewsua',
  'kpszsu',
  'veselyy_pivden',
  'mykolaivska_oda',
  'kharkivoda',
  'odessa_typical',
  'realwar_ukraine',
  'ukraine_now',
  'suspilne_news',
  'ukraine_online',
  'ukraine_radar',
  'pivden_radar',
  // --- New Channels from NEPTUN ---
  'shahedradar',
  'radarraketppo',
  'radar_raketaa',
  'radar_top_ua',
  'poltavaradar',
  'radar_dnipra',
  'eyes_everywhere_ua',
  'pivden_varta',
  'sumyregion',
  'sumygo',
  'kharkov_media',
  'odessaveter',
  'dnepr_nagladach',
  'info_zp',
  'my_safety_chernigiv'
];

const PRIVATE_TITLES = [
  'НЕПТУН',
  'Труха ⚡️ Радар'
];

// Types that represent actual target movement (shown in monitoring)
const MOVEMENT_TYPES = new Set([
  'DRONE', 'MISSILE', 'CRUISE_MISSILE', 'BALLISTIC_MISSILE', 
  'AIRCRAFT', 'KAB', 'ZIRCON', 'KH101', 'ISKANDER', 
  'KINZHAL', 'KALIBR', 'FPV', 'RECON', 'UNKNOWN'
]);

// Types to exclude from monitoring feed (but still process for map)
const EXCLUDED_FROM_MONITORING = new Set([
  'SUMMARY', 'INFO', 'ALERT', 'PPO'
]);

export async function startTelegramWorker(io: Server) {
  const apiId = Number(process.env.TELEGRAM_API_ID);
  const apiHash = process.env.TELEGRAM_API_HASH;
  const sessionString = process.env.TELEGRAM_SESSION;

  if (!apiId || !apiHash || !sessionString) {
    console.error("Telegram credentials missing. Telegram worker disabled.");
    return;
  }

  const stringSession = new StringSession(sessionString);
  const client = new TelegramClient(stringSession, apiId, apiHash, {
    connectionRetries: 5,
  });

  try {
    await client.connect();
    console.log("Connected to Telegram using string session.");

    const me = await client.getMe();
    console.log(`Logged in as: ${me.username || me.id}`);

    let source = await prisma.source.findFirst({ where: { name: 'Telegram Worker' } });
    if (!source) {
      source = await prisma.source.create({
        data: {
          name: 'Telegram Worker',
          type: 'API'
        }
      });
    }

    const sourceId = source.id;

    // ─── REALTIME HANDLER ────────────────────────────────────────────────
    client.addEventHandler(async (event: NewMessageEvent) => {
      const message = event.message;
      if (!message || !message.message) return;

      const chat = await message.getChat();
      if (!chat) return;
      
      const username = 'username' in chat && chat.username ? chat.username.toLowerCase() : null;
      const title = 'title' in chat && chat.title ? chat.title : null;
      
      const isPublicMatch = username && CHANNELS.some(c => c.toLowerCase() === username);
      const isPrivateMatch = title && PRIVATE_TITLES.some(t => title.includes(t));
      
      if (isPublicMatch || isPrivateMatch) {
        const text = message.message;
        const parsedThreats = parseTelegramText(text);
        if (!parsedThreats || parsedThreats.length === 0) return;
        
        const channelDisplay = username || title || 'Monitoring';
        const threatType = parsedThreats[0].type;

        // Only save to monitoring feed if it's actual target movement
        if (MOVEMENT_TYPES.has(threatType)) {
          const tags = [threatType];
          try {
            const savedMsg = await prisma.monitoringMessage.create({
              data: {
                text,
                channelName: channelDisplay,
                timestamp: new Date(message.date * 1000),
                tags
              }
            });
            io.emit('monitoring:new_message', savedMsg);
          } catch (e) {
            console.error("Failed to save monitoring message", e);
          }
        }

        // Add all matched targets to map (including PPO)
        for (const parsed of parsedThreats) {
          if (parsed.lat !== null && parsed.lng !== null) {
              const confidence = (parsed.confidence || 80) / 100;
              console.log(`[Parser] Detected ${parsed.type} qty=${parsed.quantity || 1} at [${parsed.lat?.toFixed(2)}, ${parsed.lng?.toFixed(2)}] conf=${confidence} from ${channelDisplay}`);
              const savedThreat = await processExternalThreat(
                  null,
                  parsed.type as any,
                  parsed.lat,
                  parsed.lng,
                  new Date(),
                  sourceId,
                  null,
                  parsed.direction,
                  confidence,
                  parsed.quantity ?? 1,
                  parsed.targetName ?? null,
                  parsed.targetLat ?? null,
                  parsed.targetLng ?? null
              );
            if (savedThreat) {
                io.emit('threat:update', savedThreat);
            }
          }
        }
      }
    }, new NewMessage({}));

    // ─── POLLING HISTORY ─────────────────────────────────────────────────
    const pollHistory = async () => {
        try {
            console.log("Fetching recent Telegram history (polling)...");
            
            const dialogs = await client.getDialogs();
            
            for (const dialog of dialogs) {
                if (!dialog.entity) continue;
                const username = 'username' in dialog.entity && dialog.entity.username ? dialog.entity.username.toLowerCase() : null;
                const title = 'title' in dialog.entity && dialog.entity.title ? dialog.entity.title : null;
                
                const isPublicMatch = username && CHANNELS.some(c => c.toLowerCase() === username);
                const isPrivateMatch = title && PRIVATE_TITLES.some(t => title.includes(t));
                
                if (isPublicMatch || isPrivateMatch) {
                    const messages = await client.getMessages(dialog.entity, { limit: 15 });
                    for (const message of messages.reverse()) {
                        if (!message || !message.message) continue;
                        const msgTime = message.date * 1000;
                        
                        const parsedThreats = parseTelegramText(message.message);
                        if (!parsedThreats || parsedThreats.length === 0) continue;
                        
                        const channelDisplay = username || title || 'Monitoring';
                        const threatType = parsedThreats[0].type;

                        // Only save to monitoring if it's actual target movement and fresh (<12h)
                        const isFreshMonitoring = (Date.now() - msgTime) < 12 * 60 * 60 * 1000;

                        if (isFreshMonitoring && MOVEMENT_TYPES.has(threatType)) {
                            const tags = [threatType];
                            try {
                                const existing = await prisma.monitoringMessage.findFirst({
                                    where: { 
                                        text: message.message,
                                        timestamp: new Date(message.date * 1000)
                                    }
                                });

                                if (!existing) {
                                    const savedMsg = await prisma.monitoringMessage.create({
                                        data: {
                                            text: message.message,
                                            channelName: channelDisplay,
                                            timestamp: new Date(message.date * 1000),
                                            tags
                                        }
                                    });
                                    io.emit('monitoring:new_message', savedMsg);
                                }
                            } catch (e) {
                                console.error('Failed to save monitoring message:', e);
                            }
                        }

                        // Spawn threats on the map if the message is recent (<2 hours)
                        const isFresh = (Date.now() - msgTime) < 2 * 60 * 60 * 1000;
                        
                        if (isFresh) {
                            for (const parsed of parsedThreats) {
                                if (parsed.lat !== null && parsed.lng !== null) {
                                    const savedThreat = await processExternalThreat(
                                        null, parsed.type as any, parsed.lat, parsed.lng,
                                        new Date(msgTime),
                                        sourceId, null, parsed.direction, parsed.confidence / 100,
                                        parsed.quantity, parsed.targetName ?? null, parsed.targetLat ?? null, parsed.targetLng ?? null
                                    );
                                    if (savedThreat) io.emit('threat:update', savedThreat);
                                }
                            }
                        }
                    }
                }
            }
            console.log("History fetch complete.");
        } catch (e) {
            console.error("History fetch error:", e);
        }
    };

    // ─── CLEANUP: Delete old messages every minute ────────────────────────
    const cleanupOldMessages = async () => {
        try {
            // Delete messages older than 6 hours
            const sixHoursAgo = new Date(Date.now() - 6 * 60 * 60 * 1000);
            const deleted = await prisma.monitoringMessage.deleteMany({
                where: { timestamp: { lt: sixHoursAgo } }
            });
            if (deleted.count > 0) {
                console.log(`[Cleanup] Deleted ${deleted.count} old monitoring messages.`);
            }

            // Also cap total messages to latest 500
            const totalCount = await prisma.monitoringMessage.count();
            if (totalCount > 500) {
                const toDelete = totalCount - 500;
                const oldest = await prisma.monitoringMessage.findMany({
                    orderBy: { timestamp: 'asc' },
                    take: toDelete,
                    select: { id: true }
                });
                await prisma.monitoringMessage.deleteMany({
                    where: { id: { in: oldest.map(m => m.id) } }
                });
                console.log(`[Cleanup] Capped messages, deleted ${toDelete} excess.`);
            }

            // Delete monitoring messages with non-movement types (cleanup legacy)
            await prisma.monitoringMessage.deleteMany({
                where: {
                    OR: [
                        { tags: { has: 'SUMMARY' } },
                        { tags: { has: 'INFO' } },
                        { tags: { has: 'ALERT' } },
                        { tags: { has: 'PPO' } },
                        { channelName: { in: ['air_alert_ua', 'ukraine_alarm_bot', 'Офіційні Тривоги'] } }
                    ]
                }
            });
        } catch (e) {
            console.error("[Cleanup] Error:", e);
        }
    };

    // Archive only threats older than 2 hours (keep recent ones alive across restarts)
    try {
        const twoHoursAgo = new Date(Date.now() - 2 * 60 * 60 * 1000);
        const archived = await prisma.threatObject.updateMany({
            where: { status: 'ACTIVE', updatedAt: { lt: twoHoursAgo } },
            data: { status: 'ARCHIVED' }
        });
        console.log(`[Startup] Archived ${archived.count} old threats (>2h).`);
    } catch(e) {}

    // Run initial cleanup
    await cleanupOldMessages();

    // Poll history once on startup to catch up on missed messages
    await pollHistory();

    // Run cleanup every minute
    setInterval(async () => {
        await cleanupOldMessages();
    }, 60 * 1000);

  } catch (error) {
    console.error("Error starting Telegram worker:", error);
  }
}

File: backend/python_parser/telegram_worker.py
import os
import asyncio
import logging
import requests
from telethon import TelegramClient, events
from telethon.sessions import StringSession
from parser import parse_telegram_text

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

api_id = os.environ.get('TELEGRAM_API_ID')
api_hash = os.environ.get('TELEGRAM_API_HASH')
session_string = os.environ.get('TELEGRAM_SESSION')

NODE_API_URL = "http://127.0.0.1:3001/api/internal/telegram-message"

CHANNELS = [
  'vanek_nikolaev', 'monitor', 'kievreal1', 'operativnoZSU', 'insiderUKR', 'smolii_ukraine', 
  'kyiv_golovne', 'war_monitor', 'monitor_ukraine', 'kyiv_vanek', 'nebo_raketa', 
  'sectorv666', 'deraketaua', 'kherson_non_drone', 'rozvidkaneba', 'kudy_letyt', 
  'nablydatel_dozor', 'krolevetsnews', 'place_kharkiv', 'ukraineradar_24_7', 'kievmap', 
  'truexanewsua', 'kpszsu', 'veselyy_pivden', 'kharkivoda', 'odessa_typical', 
  'realwar_ukraine', 'ukraine_now',
  'pivden_varta', 'radar_top_ua', 'odessa_inform', 'monitor1654', 
  'ukrainealarmsignal', 'eyes_everywhere_ua', 'vibuxaviasia', 'renihub', 'odessaveter', 
  'kharkov_media', 'pivden_fpv', 'tlknewsua', 'temporis_odesa', 'kherson_monitoring', 
  'shahedradar', 'operatyvnohlep', 'poltavaranger', 'raketa_trevoga',
  'roman_romanchuk', 'kyivoda', 'suspilne_news', 'ukraine_radar',
  'київський купол', 'ринда моніторить', 'повітряний простір', 'єрадар',
  'volynskaODA', 'zhytomyrskaODA', 'ternopilskaODA', 'khmelnytskaODA',
  'vinnytskaODA', 'chernivetskaODA', 'ifoda', 'zakarpatskaODA',
  'cherkaskaODA', 'kirovohradskaODA', 'poltavskaODA', 'sumy_oda',
  'synegubov', 'dnipropetrovskaODA', 'zoda_gov_ua', 'mykolaivskaODA',
  'odeskaODA', 'khersonskaODA', 'pavlokyrylenko_donoda', 'luhanskaVTSA', 'VA_Kyiv',
  'korabeli', 'monitorwar', 'monitoring_ukraine',
  'air_alert_ua', 'zahyst_ua', 'radar_dnepr', 'kharkiv_1654',
  'zsu_insider', 'air_alarm_ua', 'ua_drones', 'mykolaiv_online',
  'poltava_monitoring', 'cherkasy_today', 'sumy_today',
  'uaradar', 'raketna_zagroza'
]

async def main():
    if not api_id or not api_hash or not session_string:
        logger.error("Missing TELEGRAM_API_ID, TELEGRAM_API_HASH, or TELEGRAM_SESSION.")
        return

    client = TelegramClient(StringSession(session_string), int(api_id), api_hash)
    await client.connect()
    
    if not await client.is_user_authorized():
        logger.error("Session is invalid or expired. Please run login.py again.")
        return
        
    me = await client.get_me()
    logger.info(f"Logged in as {me.username or me.id}")
    
    @client.on(events.NewMessage)
    async def handler(event):
        chat = await event.get_chat()
        
        username = getattr(chat, 'username', None)
        title = getattr(chat, 'title', None)
        
        username_lower = username.lower() if username else None
        title_lower = title.lower() if title else None
        
        channels_lower = [c.lower() for c in CHANNELS]
        is_monitored = False
        
        if username_lower and username_lower in channels_lower:
            is_monitored = True
        elif title_lower and any(c in title_lower for c in channels_lower):
            is_monitored = True
            
        if not is_monitored:
            return
            
        text = event.message.message
        if not text:
            return
            
        # Handle replies
        if event.message.reply_to_msg_id:
            try:
                parent_msg = await client.get_messages(event.peer_id, ids=event.message.reply_to_msg_id)
                if parent_msg and parent_msg.message:
                    text = parent_msg.message + '\\n---\\n' + text
            except Exception as e:
                logger.error(f"Failed to fetch reply context: {e}")
                
        try:
            threats = parse_telegram_text(text)
            if not threats:
                return
                
            channel_display = username or title or 'Monitoring'
            
            payload = {
                "text": text,
                "channelName": channel_display,
                "timestamp": int(event.message.date.timestamp() * 1000),
                "threats": [t.model_dump() for t in threats]
            }
            
            res = requests.post(NODE_API_URL, json=payload)
            if res.status_code == 200:
                logger.info(f"Forwarded {len(threats)} threats from {channel_display}")
            else:
                logger.error(f"Failed to forward threats: {res.status_code} {res.text}")
            
        except Exception as e:
            logger.error(f"Error processing message: {e}")

    logger.info("Listening for new messages...")
    await client.run_until_disconnected()

if __name__ == "__main__":
    asyncio.run(main())
